### ISSUE TITLE
Linear layer weights are in float32 ?

### ISSUE THREAD

**Issue by githubsgi** [pytorch/torchtitan#1027]
### Bug description

I am seeing Linear layer weights in  float32 ( wq.weight.dtype torch.float32 )  even after setting the following. 

mixed_precision_param = "bfloat16"
mixed_precision_reduce = "bfloat16"

Is that expected or I hit upon a bug ? 

### Versions

1.  Yes. 

2. See the description section . 
3.  It is easy to check that by adding logger lines  . See below. 
```
            # Non-PP forward / backward
            with self.train_context(optional_context_parallel_ctx):
                assert len(model_parts) == 1
                logger.info(f"Linear wq.weight.dtype {model_parts[0].layers['0'].attention.wq.weight.dtype}")
                last_layer = str(len(model_parts[0].layers) -1 )
                logger.info(f"Linear wq.weight.dtype {model_parts[0].layers[last_layer].attention.wq.weight.dtype}")
                pred = model_parts[0](inputs)
                loss = self.train_spec.loss_fn(pred, labels)
                # pred.shape=(bs, seq_len, vocab_size)
                # need to free to before bwd to avoid peaking memory
                del pred
                loss.backward()
```

**Comment by tianyu-l** [comment]
Not sure about this. 
cc: @mori360 

Could you first check if the `mixed_precision_param` setting took effect? Because I saw the choice is limited to fp32:
https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L355

**Comment by mori360** [comment]
mixed_precision_param could take "bfloat16" as another choice, just added in the PR above.
Please take care that low precision at gradient, the effect from low_precision might be enlarged at gradient, thus we only keep "float32" before, but leave "bfloat16" as a choice.

**Comment by githubsgi** [comment]
Thanks. Would mean  tf32 ops instead of bf16  used in the tensorcore/systolic-arrays for  matrix multiplication  ?   That would decrease tokens/sec significantly, right ?

**Comment by awgu** [comment]
You are logging the parameter dtype outside of the module `forward()`. `param_dtype` only affects the unsharded parameter (constructed via all-gather) used for forward/backward computation, not the sharded parameter, which is in fp32 in torchtitan.

The parameters are not all-gathered until you start running `pred = model_parts[0](inputs)`.

**Comment by githubsgi** [comment]
Thanks.  Still not clear whether attentions are computed in TF32 or BF16 .   Hopefully not FP32, as that would not use the tensorcors/systolics.

**Comment by mori360** [comment]
> Still not clear whether attentions are computed in TF32 or BF16

If with mp_policy set to be BF16, them the computations will be BF16

**Comment by mori360** [comment]
> Where in toml file it is specified ?

You can add `mixed_precision_param = "bfloat16"` under [training] in the toml file

**Comment by githubsgi** [comment]
Thanks, that is already set. I guess I did not mention one more setting.   I am using compile=false .   I am seeing xq, xk, xv coming in as fp32 also.  Would not that imply  computation is happening in FP32 instead of BF16 ?

` output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)`

[at here](https://github.com/pytorch/torchtitan/blob/c7a8a5a9bbdf080a12cb0ddc9bc2a023a4d1da8c/torchtitan/models/llama/model.py#L258)

**Comment by mori360** [comment]
> Thanks, that is already set. I guess I did not mention one more setting. I am using compile=false . I am seeing xq, xk, xv coming in as fp32 also. Would not that imply computation is happening in FP32 instead of BF16 ?
> 
> ` output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)`

Could you provide more details on where you see xq, xk, xv in fp32? 
The computation here are in BF16.
After input `training.mixed_precision_param = "bfloat16"` into FSDP2, we would cast params to "bfloat16" as set during pre-forward.
So the calculation of ` output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)` as in forward is calculated in "bfloat16".

But if you see params as you first provide, they are in fp32:
assert len(model_parts) == 1
                logger.info(f"Linear wq.weight.dtype {model_parts[0].layers['0'].attention.wq.weight.dtype}")
                last_layer = str(len(model_parts[0].layers) -1 )
                logger.info(f"Linear wq.weight.dtype {model_parts[0].layers[last_layer].attention.wq.weight.dtype}")
                pred = model_parts[0](inputs)

**Comment by lkhphuc** [comment]
#630 Also are you running with more than 1 gpu, i.e using FSDP at all? Otherwise currently torchtitan runs with float32 on single gpu.

**Comment by githubsgi** [comment]
@lkhphuc , thanks. I confirmed that for 1 rank  attention self.wq, xq, and x  are fp32 .  For 2 ranks , all 3 are bf16.  What is the reason, 1 rank is fp32 ?

**Comment by githubsgi** [comment]
> You are logging the parameter dtype outside of the module `forward()`. `param_dtype` only affects the unsharded parameter (constructed via all-gather) used for forward/backward computation, not the sharded parameter, which is in fp32 in torchtitan.
> 
> The parameters are not all-gathered until you start running `pred = model_parts[0](inputs)`.

@awgu , thanks.  Would like understand more  about the fp32 and bf16  parameters.   Is there a line in code where the sharded fp32 parameters  casted to unsharded bf16 parameter ?

**Comment by mori360** [comment]
> [@lkhphuc](https://github.com/lkhphuc) , thanks. I confirmed that for 1 rank attention self.wq, xq, and x are fp32 . For 2 ranks , all 3 are bf16. What is the reason, 1 rank is fp32 ?

For 1 rank, fsdp is not applied.
https://github.com/pytorch/torchtitan/blob/ecf26c82e328916eade6720aafe2e4c7e7622e7b/torchtitan/distributed/parallel_dims.py#L116-L118
https://github.com/pytorch/torchtitan/blob/ecf26c82e328916eade6720aafe2e4c7e7622e7b/torchtitan/models/llama/parallelize_llama.py#L90-L92
